fw/http.c

```diff
@@ -6236,25 +6249,29 @@ tfw_http_msg_process_generic(TfwConn *conn, TfwStream *stream,
  * returned an error code on. The rest of skbs are freed by us.
  */
 int
-tfw_http_msg_process(TfwConn *conn, TfwFsmData *data)
+tfw_http_msg_process(TfwConn *conn, struct sk_buff *skb)
 {
 	int r = T_OK;
 	TfwStream *stream = &((TfwConn *)conn)->stream;
 	struct sk_buff *next;
 
-	if (data->skb->prev)
-		data->skb->prev->next = NULL;
-	for (next = data->skb->next; data->skb;
-	     data->skb = next, next = next ? next->next : NULL)
+	if (skb->prev)
+		skb->prev->next = NULL;
+	for (next = skb->next; skb;
+	     skb = next, next = next ? next->next : NULL)
 	{
 		if (likely(r == T_OK || r == T_POSTPONE)) {
-			data->skb->next = data->skb->prev = NULL;
-			r = TFW_CONN_H2(conn)
-				? tfw_h2_frame_process(conn, data->skb)
-				: tfw_http_msg_process_generic(conn, stream,
-							       data);
+			skb->next = skb->prev = NULL;
+			if (unlikely(TFW_CONN_PROTO(conn) & TFW_FSM_WS))
+				r = tfw_ws_msg_process(conn, skb);
```

Comment:
This looks hacky and there is no sense to call WS through HTTP FSM handler. We should just call `tfw_gfsm_register_fsm()` for `tfw_ws_msg_process()` just as we do this for this handler for the HTTP FSM.

Reply:
For now we call `tfw_gfsm_register_fsm()` for HTTP FSM only to call frang hooks. But `tfw_http_msg_process()` called directly from ss_hooks.

So if you prefer full separation of websocket protocol from `tfw_http_msg_process()` than we need separate websocket ss_hooks.

Reply:
I have factored out common code for recv in `tfw_connection_recv()` in `connection.c`. These (websocket and http processing callbacks) are now separate.